storage/aws/antispam/aws.go

```diff
@@ -0,0 +1,437 @@
+
+		}
+		if e.curRI.First > 0 {
+			e.curData = e.curData[e.curRI.First:]
+		}
+		if len(e.curData) > int(e.curRI.N) {
+			e.curData = e.curData[:e.curRI.N]
+		}
+		e.i = 0
+	}
+	t, e.curData = e.curData[0], e.curData[1:]
+	rIdx := e.curRI.Index*layout.EntryBundleWidth + uint64(e.curRI.First) + e.i
+	e.i++
+	return rIdx, t, nil
+}
+
+// follower is a struct which knows how to populate the antispam storage with identity hashes
+// for entries in a log.
+type follower struct {
+	as           *AntispamStorage
+	bundleHasher func([]byte) ([][]byte, error)
+}
+
+func (f *follower) Name() string {
+	return "AWS antispam"
+}
+
+// Follow uses entry data from the log to populate the antispam storage.
+func (f *follower) Follow(ctx context.Context, lr tessera.LogReader) {
+	errOutOfSync := errors.New("out-of-sync")
+
+	t := time.NewTicker(time.Second)
+	var (
+		entryReader *entryStreamReader[[]byte]
+		stop        func()
+
+		curEntries [][]byte
+		curIndex   uint64
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+		}
+		size, err := lr.IntegratedSize(ctx)
+		if err != nil {
+			klog.Errorf("Populate: IntegratedSize(): %v", err)
+			continue
+		}
+
+		// Busy loop while there's work to be done
+		for workDone := true; workDone; {
+			err = func() error {
+				tx, err := f.as.dbPool.BeginTx(ctx, &sql.TxOptions{
+					ReadOnly: false,
+				})
+				if err != nil {
+					return err
+				}
+
+				row := tx.QueryRowContext(ctx, "SELECT nextIdx FROM AntispamFollowCoord WHERE id = 0")
+
+				var followFrom uint64
+				if err := row.Scan(&followFrom); err != nil {
+					return err
+				}
+
+				if followFrom >= size {
+					// Our view of the log is out of date, exit the busy loop and refresh it.
+					workDone = false
+					return nil
+				}
+
+				f.as.pushBack.Store(size-followFrom > uint64(f.as.opts.PushbackThreshold))
+
+				// If this is the first time around the loop we need to start the stream of entries now that we know where we want to
+				// start reading from:
+				if entryReader == nil {
+					next, st := lr.StreamEntries(ctx, followFrom)
+					stop = st
+					entryReader = newEntryStreamReader(next, f.bundleHasher)
+				}
+
+				bs := uint64(f.as.opts.MaxBatchSize)
+				if r := size - followFrom; r < bs {
+					bs = r
+				}
+				batch := make([][]byte, 0, bs)
+				for i := range int(bs) {
+					idx, c, err := entryReader.Next()
+					if err != nil {
+						return fmt.Errorf("entryReader.next: %v", err)
+					}
+					if wantIdx := followFrom + uint64(i); idx != wantIdx {
+						// We're out of sync
+						return errOutOfSync
+					}
+					batch = append(batch, c)
+				}
+				curEntries = batch
+				curIndex = followFrom
+
+				// Store antispam entries.
+				//
+				// Note that we're writing the antispam entries outside of the transaction here. The reason is because we absolutely do not want
+				// the transaction to fail if there's already an entry for the same hash in the IDSeq table.
+				//
```

Comment:
You don't have this problem in MySQL since you're using `INSERT IGNORE`, I wonder if it might be best to just insert these rows in the `tx` you have open for now and then look at other approaches if there's a performance optimsation which could be done, etc.?

Reply:
Thanks for confirming this. This was something I wanted to talk to you about, but forgot because of the break between starting this PR and finishing it.